File: 2015/cpp/day08.cpp
#include <cassert>
#include <string>
#include <vector>
#include <fstream>

using namespace std;

auto Problem1(const vector<string> &input) -> int {
    int ret = 0;
    for (const auto &s : input) {
        int len = s.size();
        int i = 0;
        int chars = -2; // two double quotes
        while (i < len) {
            switch (s[i]) {
            case '\\':
                ++i;
                switch (s[i]) {
                case '\\':
                    ++i;
                    break;
                case '"':
                    ++i;
                    break;
                case 'x':
                    i += 3;
                    break;
                }
                break;
            default:
                ++i;
                break;
            }
            ++chars;
        }

        ret += len - chars;
    }
    return ret;
}

auto Problem2(const vector<string> &input) -> int {
    int ret = 0;
    for (const auto &s : input) {
        int len = s.size();
        int chars = 2; // two double quotes
        for (char c : s) {
            switch (c) {
            case '"':
            case '\\':
                chars += 2;
                break;
            default:
                chars += 1;
                break;
            }
        }

        ret += chars - len;
    }
    return ret;
}

auto Test() -> void {
    vector<string> input{
        "\"\"",
        "\"abc\"",
        "\"aaa\\\"aaa\"",
        "\"\\x27\"",
    };

    int ret1 = Problem1(input);
    assert(ret1 == 12);

    int ret2 = Problem2(input);
    assert(ret2 == 19);
}

int main() {
    Test();

    ifstream f("../input/day08.txt");
    vector<string> input;
    string tmp;
    while (getline(f, tmp)) {
        input.push_back(tmp);
    }

    int ret1 = Problem1(input);
    assert(ret1 == 1371);

    int ret2 = Problem2(input);
    assert(ret2 == 2117);

    printf("ret1 = %d\n", ret1);
    printf("ret2 = %d\n", ret2);
    return 0;
}

File: 2015/python/day08.py
#!/usr/bin/env python3


def count_real_chars(s: str) -> int:
    s = s[1 : len(s) - 1]
    i = 0
    ret = 0
    while i < len(s):
        match s[i]:
            case "\\":
                i += 1
                match s[i]:
                    case '"' | "\\":
                        i += 1
                        ret += 1
                    case "x":
                        i += 3
                        ret += 1
            case _:
                i += 1
                ret += 1

    return ret


def problem1(lines: list[str]) -> int:
    return sum([len(s) - count_real_chars(s) for s in lines])


def count_encoded_chars(s: str) -> int:
    ret = 2
    for c in s:
        match c:
            case '"' | "\\":
                ret += 2
            case _:
                ret += 1
    return ret


def problem2(lines: list[str]) -> int:
    return sum([count_encoded_chars(s) - len(s) for s in lines])


with open("../input/day08.txt") as f:
    lines = f.readlines()

    ret1 = problem1(lines)
    assert ret1 == 1371

    ret2 = problem2(lines)
    assert ret2 == 2117

    print(f"ret1={ret1}")
    print(f"ret2={ret2}")
